subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -0,0 +1,338 @@
+   * into, for example, excel..  note: it would be nice to show
+   * timestamp too.. need to check if there is a sane way to log
+   * timestamp of last rendered buffer, so we could correlate dips
+   * in framerate to certain positions in the stream.
+   */
+  if (dr == 0.0) {
+    gst_tracer_record_log (tr_framerate, self->pid, current_ts, frames_rendered,
+        frames_dropped, average_fps, -1.0f, -1.0f);
+  } else {
+    gst_tracer_record_log (tr_framerate, self->pid, current_ts, frames_rendered,
+        frames_dropped, -1.0f, rr, dr);
+  }
+
+  self->last_frames_rendered = frames_rendered;
+  self->last_frames_dropped = frames_dropped;
+  self->last_ts = current_ts;
+
+  return TRUE;
+}
+
+static void
+gst_framerate_tracer_calculate_framerate (GstFramerateTracer *self,
+    GstClockTime ts)
+{
+  g_atomic_int_inc (&self->frames_rendered);
+
+  if (G_UNLIKELY (!GST_CLOCK_TIME_IS_VALID (self->start_ts))) {
+    self->interval_ts = self->last_ts = self->start_ts = ts;
+  }
+  if (GST_CLOCK_DIFF (self->interval_ts, ts) > self->fps_update_interval) {
+    gst_framerate_tracer_display_current_fps (self, ts);
+    self->interval_ts = ts;
+  }
+}
+
+/* FIXME: Undefined behavior if multiple video sinks exist. */
+gboolean
+is_target_video_sink (GstElement *element)
+{
+  GstElement *parent;
+  const gchar *klass;
+
+  if (!GST_IS_ELEMENT (element))
+    return FALSE;
+
+  klass = gst_element_class_get_metadata (GST_ELEMENT_GET_CLASS (element),
+      GST_ELEMENT_METADATA_KLASS);
+  parent = GST_ELEMENT_PARENT (element);
+
+  return GST_OBJECT_FLAG_IS_SET (element, GST_ELEMENT_FLAG_SINK) &&
+      klass &&
+      g_strstr_len (klass, -1, "Video") && g_strstr_len (klass, -1, "Sink");
+}
+
+static void
+do_push_buffer_pre (GstTracer *tracer, GstClockTime ts, GstPad *pad,
+    GstBuffer *buf)
+{
+  GstFramerateTracer *self = (GstFramerateTracer *) tracer;
+  GstPad *peer_pad;
+  GstElement *element;
+
+  peer_pad = get_peer_pad (pad);
+  if (!peer_pad)
+    return;
+
+  element = gst_pad_get_parent_element (peer_pad);
+  if (!is_target_video_sink (element))
+    goto beach;
+
+  gst_framerate_tracer_calculate_framerate (tracer, ts);
+
+beach:
+  if (peer_pad)
+    gst_object_unref (peer_pad);
+}
+
+/* tracer class */
+
+static void
+gst_framerate_tracer_constructed (GObject *object)
+{
+  GstFramerateTracer *self = GST_FRAMERATE_TRACER (object);
+  gchar *params, *tmp;
+  GstStructure *params_struct = NULL;
+
+  g_object_get (self, "params", &params, NULL);
+
+  if (!params)
+    return;
+
+  tmp = g_strdup_printf ("framerate,%s", params);
+  params_struct = gst_structure_from_string (tmp, NULL);
+  g_free (tmp);
+
+  if (params_struct) {
+    const gchar *name, *flags;
+    /* Set the name if assigned */
+    name = gst_structure_get_string (params_struct, "name");
+    if (name)
+      gst_object_set_name (GST_OBJECT (self), name);
+
+    /* Read the flags if available */
+    flags = gst_structure_get_string (params_struct, "flags");
```

Comment:
we get the flags string, but not use it anyhow.

YAGNI

Reply:
I will use it, anyway...

Reply:
ok